### PR TITLE
Add content rw s3

### PR DESCRIPTION
**Pull request by davidbalazs93**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,148 @@
...
+## Installation
+
+1. go get github.com/kardianos/govendor
+2. go get github.com/Financial-Times/content-rw-s3
+3. cd $GOPATH/src/github.com/Financial-Times/content-rw-s3
+4. govendor sync
+
+
+## Running locally
+
```
Comment (tosan88) [inline README.md]: There are some bits which seem outdated below

File: README.md [hunk 2]
```diff
@@ -1,2 +1,148 @@
...
+### DELETE /UUID
+Will return 204
+
+## Utility endpoints
+
+### GET /
+Streams all payloads in a given bucket
+
+### GET /__ids
```
Comment (tosan88) [inline README.md]: Probably a leftover

File: helm/content-rw-s3/app-configs/content-rw-s3_delivery.yaml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# Values used for the deployed application.
+replicaCount: 1
+service:
+  name: content-rw-s3
+bucket:
```
Comment (tosan88) [inline helm/content-rw-s3/app-configs/content-rw-s3_delivery.yaml]: I guess this is not needed

File: main.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	wh := service.NewWriterHandler(w, r)
+	rh := service.NewReaderHandler(r)
+
+	servicesRouter := mux.NewRouter()
+	service.Handlers(servicesRouter, wh, rh, resourcePath)
+	service.AddAdminHandlers(servicesRouter, svc, bucketName)
+
+	log.Infof("listening on %v", port)
+
+	if err := http.ListenAndServe(":" + port, nil); err != nil {
```
Comment (tosan88) [inline main.go]: Hm, strange, I would expected to be the `servicesRouter` as the second parameter...

File: service/model.go [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package service
+
+type obj struct {
```
Comment (tosan88) [inline service/model.go]: It seems unused

File: service/processor.go [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+	}
+
+	return "", false, nil
+}
+
+func (r *S3Reader) listObjects(keys chan <- *string, uuid string) error {
+	return r.svc.ListObjectsV2Pages(r.getListObjectsV2Input(uuid),
+		func(page *s3.ListObjectsV2Output, lastPage bool) bool {
+			for _, o := range page.Contents {
+				if (!strings.HasSuffix(*o.Key, "/") && !strings.HasPrefix(*o.Key, "__")) && (*o.Key != ".") {
```
Comment (tosan88) [inline service/processor.go]: are the "__" and "." checks AWS specific?

File: service/processor.go [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+	return ReaderHandler{reader: reader}
+}
+
+type ReaderHandler struct {
+	reader Reader
+}
+
+func (rh *ReaderHandler) HandleGet(rw http.ResponseWriter, r *http.Request) {
+	uuid := uuid(r.URL.Path)
+	publishedDate := r.URL.Query().Get("publishedDate")
```
Comment (tosan88) [inline service/processor.go]: Could the param name be just `date` as for write?

File: service/processor_test.go [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+	"net/http/httptest"
+	"strconv"
+	"strings"
+	"sync"
+	"testing"
+)
+
+const (
+	expectedUUID = "123e4567-e89b-12d3-a456-426655440000"
+	expectedMessageID = "7654e321-b98e-3d12-654a-000042665544"
```
Comment (tosan88) [inline service/processor_test.go]: It is not used

File: service/processor_test.go [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+	s.payload = "ERROR PAYLOAD"
+	b, i, ct, err := r.Get(expectedUUID, "")
+	assert.Error(t, err)
+	assert.Equal(t, s.s3error, err)
+	assert.False(t, b)
+	assert.Nil(t, i)
+	assert.Nil(t, ct)
+}
+
+func generateKeys(count int, addIgnoredKeys bool) s3.ListObjectsV2Output {
```
Comment (tosan88) [inline service/processor_test.go]: Is this used?

File: vendor/vendor.json [hunk 10]
```diff
@@ -0,0 +1,377 @@
+{
+	"comment": "",
+	"ignore": "test",
+	"package": [
+		{
+			"path": "appengine",
```
Comment (tosan88) [inline vendor/vendor.json]: What does this appengine do?

File: main.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+	logMetrics := app.Bool(cli.BoolOpt{
+		Name:   "logMetrics",
+		Value:  false,
+		Desc:   "Whether to log metrics. Set to true if running locally and you want metrics output",
+		EnvVar: "LOG_METRICS",
+	})
+
+	app.Action = func() {
+
```
Comment (ioanaciont) [inline main.go]: remove empty lines

File: service/handlers_test.go [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+	sync.Mutex
+	found       bool
+	uuid        string
+	payload     string
+	rc          io.ReadCloser
+	returnError error
+	returnCT    string
+}
+
+func (r *mockReader)GetPublishDateForUUID(uuid string) (string, bool, error) {
```
Comment (ioanaciont) [inline service/handlers_test.go]: Hmm, is it not formatted correctly?

File: service/processor.go [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+		rw.Header().Set("Content-Type", "application/json")
+		rw.WriteHeader(http.StatusBadRequest)
+		rw.Write([]byte("{\"message\":\"Required query param 'date' was not provided.\"}"))
+		return
+	}
+
+	rw.Header().Set("Content-Type", "application/json")
+	var err error
+	bs, err := ioutil.ReadAll(r.Body)
+
```
Comment (ioanaciont) [inline service/processor.go]: empty line here is not necessary

**Pull request merged**